### ISSUE TITLE
AcquireTokenByDeviceCodeAsync cannot acquire jwt tokens with "aud": "spn:xxx" prefix

### ISSUE THREAD

**Issue by lodejard** [AzureAD/azure-activedirectory-library-for-dotnet#1493]
**Which Version of ADAL are you using ?**
ADAL 4.5.0

**Which platform has the issue?**
netcore

**What authentication flow has the issue?**
* Desktop / Mobile
    * [ ] Interactive
    * [ ] Integrated Windows Auth
    * [ ] Username Password
    * [x] Device code flow (browserless)
* Web App
    * [ ] Authorization code
    * [ ] OBO
* Web API
    * [ ] OBO

Other? - please describe;

**Is this a new or existing app?**
b. The app is in production, uses the same version of ADAL, but started seeing this issue

**Repro**

```csharp
try
{
    result = await ctx.AcquireTokenSilentAsync(resourceId, clientId);
}
catch
{
    var codeResult = await ctx.AcquireDeviceCodeAsync(resourceId, clientId, "api-version=1.0");
    _console.Error.WriteLine(codeResult.Message.Color(ConsoleColor.DarkYellow));
    result = await ctx.AcquireTokenByDeviceCodeAsync(codeResult);
}
```

**Expected behavior**
A jwt access token with the claim `"aud": "spn:<webapp-guid>"` 

**Actual behavior**
A jwt access token with the claim `"aud": "<webapp-guid>"`

**Possible Solution**
Add an overload to AcquireTokenByDeviceCodeAsync which also accepts `extraQueryParameters`.

**Additional context/ Logs / Screenshots**
`kubectl` uses autorest/adal-for-go, and appears to put api-version=1.0 on both device-code calls. I'm writing a c# client app, but the tokens I acquire do not have the `spn:` prefix.

I believe the difference is because the c# adal is not adding the api-version to the second call, so the authority isn't returning the legacy prefix.

Unfortunately the kubernetes server validates the aud with the `spn:` prefix, so my c# client app is unable to provide an acceptable token if it cannot access the 1.0 auth flow.

token claims returned to kubectl
``` json
{
  "aud": "spn:bcac2066-6172-4192-ad37-7cc71317f84a",
  "iss": "https://sts.windows.net/553f0ec4-e086-4080-b811-059f2482551e/",
  "iat": 1549404507,
  "nbf": 1549404507,
  "exp": 1549408407,
  "acr": "1",
```

token claims returned to c# app
``` json
{
  "aud": "bcac2066-6172-4192-ad37-7cc71317f84a",
  "iss": "https://sts.windows.net/553f0ec4-e086-4080-b811-059f2482551e/",
  "iat": 1549404651,
  "nbf": 1549404651,
  "exp": 1549408551,
  "acr": "1",
```

**Comment by jmprieur** [comment]
@lodejard : To unblock you (and this is a hack), you could set (even programmatically and locally using the Environment class in C#) the `ExtraQueryParameter` environment variable.

Now to really fix the problem
- I'd like to be sure I understand: are you writing that you've observed a recent a regression with the same version of the library (that is a regression on the Azure AD side?)
- We would not provide an override in ADAL.NET. We'd advise you to move to MSAL.NET (in a couple of week, we should release MSAL.NET 3.0 which enables you to specify overrides for every flow).

**Comment by lodejard** [comment]
Thanks @jmprieur I'll give that a try. That env var is a very scary feature, but if it works I guess I'm glad it's there.

* I'd like to be sure I understand: are you writing that you've observed a recent a regression with the same version of the library (that is a regression on the Azure AD side?)

It was hard to choose the right answer for the "new or existing app" question. This is existing app which can send REST API calls generically to AAD authenticated resources. So although the c# app, ADAL library, and Azure AD have not changed - there is a new scenario which surfaced the problem: using the app to invoke Kubernetes cluster REST API. 

A cluster can be configured to accept AAD authn, but the cluster's server and client (kubectl) are written in `go` and both use `api-server=1.0` - and the server appears to have baked-in knowledge of the spn prefix. So the new requirement for the c# app is to be able to acquire an access token in a way that's identical to kubectl. ( see also https://github.com/kubernetes/kubernetes/blob/d210b4bcf350308f51d8bffb7748b2cc3fb611c0/vendor/github.com/Azure/go-autorest/autorest/adal/config.go#L61 )

* We would not provide an override in ADAL.NET. We'd advise you to move to MSAL.NET (in a couple of week, we should release MSAL.NET 3.0 which enables you to specify overrides for every flow).

I took a quick look at that - I guess I'd have two questions about that move.

1. Would we be able to make the api-version=1.0 calls to get spn prefixed claims on the v2 url?

2. Right now the a rest api declares the `{tenant, resource, client}` as the parameters of an auth context. for a request. It appears that MSAL is shifting to a scopes array instead of a resource identifier. Is it possible to use the resource identifier as a scope? Or when acquiring a token with the new library will the c# app need to know what scopes to request to make the call? 

In other words, will we need to change the declaration from `{tenant, resource, client}` to `{tenant, scopes, client}`?

**Comment by jmprieur** [comment]
@lodejard 
the api-version question is a good question. I don't know. 
Indeed, you would have to transform the resource into scopes. See https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/wiki/Adal-to-Msal#scopes-for-a-web-api-accepting-v10-tokens.

Do you think we can close this issue?

**Comment by lodejard** [comment]
Yep, can close.

Also, update on this, it turns out the v2 url does work for kubernetes authentication (using 'oidc' auth provider instead of 'azure' auth provider).

One last question. It looks like the new v2 endpoint is providing groups claims incorrectly when there are so many they are provided via url reference. That's a server-side bug. Do you know the best place to report that?

**Comment by jmprieur** [comment]
This is not a bug. this is by design
when a user belongs to too many groups, the size of the token would be bigger than the limit size for URLs or HTTP headers. Therefore Azure AD provides a URL to fetch the group.

you can find more information about this here: https://github.com/Azure-Samples/active-directory-dotnet-webapp-groupclaims#groups-overage-claim

**Comment by omerlh** [comment]
Hey @lodejard - can clarify on:
> Also, update on this, it turns out the v2 url does work for kubernetes authentication (using 'oidc' auth provider instead of 'azure' auth provider).
Can you please share a working code sample? I'm trying to write a code that request a token for Kuberentes API.

**Comment by jmprieur** [comment]
@omerlh : what is an auth provider? in which context do you use them. Can you please share some code or link illustrating it ? I suspect that ADAL is called by some other library? (I'm not familiar with the Kubernetes API. Does it provide the notion of auth provider?)

**Comment by omerlh** [comment]
I want to authorize users to Kubernetes dashboard. So, I want to build a small web application that redirects users to log in, acquire a token for the Kubernetes app and proxy requests to the dashboard using this token. The main missing point is how to request a token that will be valid for the Kubernetes server (because of the `spn`). See this AKS engine issues - Azure/aks-engine#376

**Comment by jmprieur** [comment]
Thanks for the context. Do you know if there is a reason why the AKS engine requires an obsolete token format? maybe that's what should be fixed (here we are only trying to work around things)

**Comment by omerlh** [comment]
I have no idea - you can look on the issue, it's open since November and no response yet. I also pinged Azure support at the same time - no response yet... Can you please responsed on the issue and clarify that this token format is obsolete?

**Comment by jmprieur** [comment]
@omerlh I'll try to find someone from that team and understand what's going on.
cc: @henrik-me

**Comment by omerlh** [comment]
Thanks! I hope this will be resolved soon 🙏

**Comment by lodejard** [comment]
@omerlh Sorry for the delay. I don't have a clean example --- but if you're using aks-engine to create your cluster something I've seen which works is to configure the `--oidc switches` on the api server. You'd do that instead of providing the information in the aapProfile tenant.

```
kubernetesConfig:
  apiServerConfig:
    --oidc-client-id: {{ the-apiserver-appid }}
    --oidc-issuer-url: https://login.microsoftonline.com/{{ the-aad-tenant-guid }}/v2.0
    --oidc-username-claim: oid
    --oidc-groups-claim: groups
```

Basically look at the properties of the jwt token your web app is passing to k8s with https://jwt.io/ --- the `"aud"` must match the `--oidc-client-id` and the `"iss"` must match the `--oidc-issuer-url`

Biggest down-side is that I haven't seen a way to get a good `kubectl` authentication experience... You can use the built-in `oidc` user credentials provider --- but the `oidc` provider seems to expect to see `id_token` and `refresh_token` already in the kubeconfig already, and doesn't seem to provide an interactive command-line login experience when those are not present yet... 

https://github.com/int128/kubelogin can kind of help there, but at that point the solution is starting to feel kind of stitched together

**Comment by lodejard** [comment]
Also @jmprieur I don't know if there's a specific reason the AzureAD login for kubernetes chose to use the older api-version which has the spn prefix. It might have been as simple as that it was the default behavior of the go-autorest client library...

Here's a line from the `kubectl` client library which is using adal device code flow to acquire a token for the cluster apiserver.

https://github.com/kubernetes/kubernetes/blob/092f3988255801ce7a97de5448384c50c400a3a4/staging/src/k8s.io/client-go/plugin/pkg/client/auth/azure/azure.go#L347

And that method seems to be calling this one, `NewOAuthConfig` from the go-autoreset adal dependency, which appears to default to adding `api-version=1.0` onto the AAD http calls...

https://github.com/kubernetes/kubernetes/blob/092f3988255801ce7a97de5448384c50c400a3a4/vendor/github.com/Azure/go-autorest/autorest/adal/config.go#L44

From what I've seen that seems to be the what's causing the `spn` prefix. The fact that the server is configured to expect the prefix might be simply that it's what the client was ending up with.

**Comment by omerlh** [comment]
Thanks @lodejard for your detailed response! I know I can change the oidc settings on the server, but as you mentioned - this will break kubectl authentication. I guess the first step will be to add support in kubectl to a token without the `spn`. I can help, but I assume this will require someone from AKS team to make it actually happen.

**Comment by jmprieur** [comment]
See also https://github.com/Azure/aks-engine/issues/376
